RFC-0004-functions-stats.md

```diff
@@ -0,0 +1,420 @@
+}
+
+```
+
+`StatsPropagationBehavior` is an enum with following options to control how stats are handled.
+
+```java
+
+public enum StatsPropagationBehavior
+{
+    /** Use the max value across all arguments to derive the new stats value */
+    USE_MAX_ARGUMENT,
+    /** Sum the stats value of all arguments to derive the new stats value */
+    SUM_ARGUMENTS,
+    /** Propagate the source stats as-is */
+    USE_SOURCE_STATS,
+    /** Use the value of output row count. */
+    ROW_COUNT,
+    /** Use the value of row_count * (1 - null_fraction). */
+    ROW_COUNT_TIMES_INV_NULL_FRACTION,
+    /** use the value of TYPE_WIDTH in varchar(TYPE_WIDTH) */
+    USE_TYPE_WIDTH_VARCHAR,
+    /** Take max of type width of arguments with varchar type. */
+    MAX_TYPE_WIDTH_VARCHAR,
+    /** Stats are unknown and thus no action is performed. */
+    UNKNOWN
+}
+```
+
+Examples of using above annotations for various function in `StringFunctions.java`
+
+1. upper(Slice)
+```java
+    @Description("converts the string to upper case")
+    @ScalarFunction
+    @LiteralParameters("x")
+    @SqlType("varchar(x)")
+    public static Slice upper(@ScalarPropagateSourceStats(propagateAllStats = true) @SqlType("varchar(x)") Slice slice)
+    {
+        return toUpperCase(slice);
+    }
+```
+
+2. starts_with(Slice, Slice): an example of constant stats.
+```java
+    @SqlNullable
+    @Description("Returns whether the first string starts with the second")
+    @ScalarFunction("starts_with")
+    @LiteralParameters({"x", "y"})
+    @SqlType(StandardTypes.BOOLEAN)
+    @ScalarFunctionConstantStats(distinctValuesCount = 2, nullFraction = 0, minValue = 0, maxValue = 1)
+    public static Boolean startsWith(@SqlType("varchar(x)") Slice x, @SqlType("varchar(y)") Slice y)
+    {
+        if (x.length() < y.length()) {
+            return false;
+        }
+
+        return x.equals(0, y.length(), y, 0, y.length());
+    }
+```
+
+3. concat(Slice, Slice): 
+
+```java
+   @Description("concatenates given character strings")
+    @ScalarFunction
+    @LiteralParameters({"x", "y", "u"})
+    @Constraint(variable = "u", expression = "x + y")
+    @SqlType("char(u)")
+    public static Slice concat(@LiteralParameter("x") Long x,
+            @ScalarPropagateSourceStats(propagateAllStats = true, nullFraction = USE_MAX_ARGUMENT, avgRowSize = SUM_ARGUMENTS, distinctValuesCount = ROW_COUNT) @SqlType("char(x)") Slice left,
+            @SqlType("char(y)") Slice right)
+    { //...
+}
+```
+ __Propagate all stats but, for nullFraction, avgRowSize and distinctValuesCount follow as specified.__
+
+4. An example of using both `ScalarFunctionConstantStats` and `ScalarPropagateSourceStats`.
+```java
+    @ScalarFunctionConstantStats(minValue = 0)
+    public static long levenshteinDistance(
+            @ScalarPropagateSourceStats(propagateAllStats = true, maxValue = MAX_TYPE_WIDTH_VARCHAR, distinctValuesCount = MAX_TYPE_WIDTH_VARCHAR) @SqlType("varchar(x)") Slice left,
```

Comment:
Are there many cases where `propagateAllStats` will be false when this annotation is specified? Should we default it to `true` so that it doesn't have to be specified?

Reply:
There a good number of them, but this is something I am going to keep in mind. If there are more functions with propagateAllStats = true, then make it the default value.